### PR TITLE
migrate from_fedora contrib spec to mapping name spec, with additional normalizations

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

We want to migrate the from_fedora and to_fedora mapping specs to a land where all the mapping specs are above average

spec/services/cocina/from_fedora/descriptive/contributor_spec.rb  contents were moved to spec/services/cocina/mappings/mods/name_spec.rb, or were found to be duplicates.

## How was this change tested?

mods normalization changed and we got a 0.006% improvement in our success rate!   Go me!

### this branch

```
Status (n=50000):
  Success:   49117 (98.234%)
  Different: 571 (1.142%)
  To Cocina error:     1 (0.002%)
  To Fedora error:     0 (0.0%)
  Missing:     311 (0.622%)
```

### main branch

```
Status (n=50000):
  Success:   49114 (98.228%)
  Different: 574 (1.148%)
  To Cocina error:     1 (0.002%)
  To Fedora error:     0 (0.0%)
  Missing:     311 (0.622%)
```

### CODE REVIEW

File: app/services/cocina/mods_normalizers/name_normalizer.rb [hunk 1]
```diff
@@ -56,6 +62,38 @@ def normalize_dupes_for(base_node)
...
+
+      def normalize_type
+        ng_xml.root.xpath('//mods:name[(@type)]', mods: ModsNormalizer::MODS_NS).each do |name_node_w_type|
+          raw_type = name_node_w_type['type']
+          next if FromFedora::Descriptive::Contributor::ROLES.keys.include?(raw_type)
+
+          if FromFedora::Descriptive::Contributor::ROLES.keys.include?(raw_type.downcase)
+            name_node_w_type['type'] = raw_type.downcase
+          else
+            name_node_w_type.remove_attribute('type')
```
Comment (justinlittman) [inline app/services/cocina/mods_normalizers/name_normalizer.rb:74]: Not a big deal, but elsewhere we've been using `.delete()`.

File: app/services/cocina/mods_normalizers/name_normalizer.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require_relative '../from_fedora/descriptive/contributor'
```
Comment (justinlittman) [inline app/services/cocina/mods_normalizers/name_normalizer.rb:3]: I'm surprised that this is required.

File: app/services/cocina/mods_normalizers/name_normalizer.rb [hunk 3]
```diff
@@ -56,6 +62,38 @@ def normalize_dupes_for(base_node)
           dupe_name_nodes[1..].each(&:remove)
         end
       end
+
+      def normalize_type
+        ng_xml.root.xpath('//mods:name[(@type)]', mods: ModsNormalizer::MODS_NS).each do |name_node_w_type|
```
Comment (justinlittman) [inline app/services/cocina/mods_normalizers/name_normalizer.rb:67]: I don't think the () are needed in "(@type)".

File: spec/services/cocina/mapping/descriptive/mods/name_spec.rb [hunk 4]
```diff
@@ -1919,4 +1919,480 @@
...
+              }
+            ]
+          }
+        end
+
+        let(:warnings) { [Notification.new(msg: 'Name type incorrectly capitalized', context: { type: 'Personal' })] }
+      end
+    end
+
+    context 'when unrecognized' do
```
Comment (justinlittman) [inline spec/services/cocina/mapping/descriptive/mods/name_spec.rb:1965]: When what is unrecognized?

File: spec/services/cocina/mapping/descriptive/mods/name_spec.rb [hunk 5]
```diff
@@ -1919,4 +1919,480 @@
...
+        let(:warnings) do
+          [
+            Notification.new(msg: 'Name/namePart type attribute set to ""'),
+            Notification.new(msg: 'name/role/roleTerm missing value')
+          ]
+        end
+      end
+    end
+
+    context 'with a role that has no URI and has xlink uris from MODS 3.3' do
```
Comment (justinlittman) [inline spec/services/cocina/mapping/descriptive/mods/name_spec.rb:2219]: Why not use the shared example? Might require some changes to allow setting the mods header.

**Review by justinlittman — CHANGES_REQUESTED** [review]
Mostly suggestions.

File: app/services/cocina/mods_normalizers/name_normalizer.rb (hunk 3 above)
Comment (ndushay) [reply app/services/cocina/mods_normalizers/name_normalizer.rb:67]: right you are.

File: spec/services/cocina/mapping/descriptive/mods/name_spec.rb (hunk 5 above)
Comment (ndushay) [reply spec/services/cocina/mapping/descriptive/mods/name_spec.rb:2219]: i felt the path of least resistance was to just port the MODS 3.3 example as it was.  If we want to make it possible to use different MODS versions, we should maybe write a ticket for that;  I don't think we have more than a handful of these tests.

File: app/services/cocina/mods_normalizers/name_normalizer.rb (hunk 2 above)
Comment (ndushay) [reply app/services/cocina/mods_normalizers/name_normalizer.rb:3]: well, I'll be.  it isn't required now.  It was when I was developing the code;  must've had something weird going on for a spell.

File: app/services/cocina/mods_normalizers/name_normalizer.rb (hunk 1 above)
Comment (ndushay) [reply app/services/cocina/mods_normalizers/name_normalizer.rb:74]: ja - docs imply they are aliases for one another.  Leaving this unless it really bugs you.

File: spec/services/cocina/mapping/descriptive/mods/name_spec.rb (hunk 4 above)
Comment (ndushay) [reply spec/services/cocina/mapping/descriptive/mods/name_spec.rb:1965]: amended.

File: spec/services/cocina/mapping/descriptive/mods/name_spec.rb (hunk 5 above)
Comment (ndushay) [reply spec/services/cocina/mapping/descriptive/mods/name_spec.rb:2219]: Also wasn't sure if we had to have a specific xlink namespace declaration on the outer MODS element.

**Comment by ndushay** [comment]
nice - elegant.  It just wasn't clicking for me.

**Review by justinlittman — APPROVED** [review]

**Pull request merged**